### ISSUE TITLE
Sometimes doesn't stop after 5 minutes despite "Run according to time schedule" ticked

### ISSUE THREAD

**Issue by JohnVeness** [Catfriend1/syncthing-android#803]
I have "Run according to time schedule" ticked and usually it works, but sometimes I notice that the notification is still showing for several hours or even days, and that the "uptime" is showing this too. Doing Force Stop then Follow Run Conditions seems to temporarily fix.
![](https://i.imgur.com/MqWzigs.png)
![](https://i.imgur.com/WNuZWwf.png)

**Comment by Catfriend1** [comment]
Hi, interesting observation. I doubt everybody has this, but verbose logs (app settings - troubleshooting - enable verbose logs) and maybe for easier read normal logs could shed some light on what's happening. I see the VPN key and know the app sends a request to shutdown Syncthing to the native binary via REST Api. Does it work correctly if you temporarily disable some of your other tools and test ?

**Comment by JohnVeness** [comment]
Ah, good spot on the VPN icon. I use AdAway in (non-root-required) VPN mode, and I wonder if that is related, as I only started using AdAway in recent week and only noticed the problem with Syncthing in recent weeks! But it doesn't happen all the time so it's not a simple cause and effect. I will do some more testing and log hunting and let you know.

**Comment by JohnVeness** [comment]
An update: I disabled the VPN and the problem seems to have reduced in frequency (used to happen every day), but it still happens once or twice a week or so. Sometimes I notice it in the morning, when it says it's been running for four hours or so. And I noticed it once when I was returning from a car journey, during which I had no Internet access (I turn off mobile data most of the time). Not that it should need Internet to access the local REST API, but I wonder if that's related?

I notice that when it says it is still running, it really is - that is, I can check on my other devices and see that they are connected to the phone, and if I copy a file I see the change at the other end. So it's not just a stuck notification icon.

Logs don't appear to be useful, even with verbose mode enabled. When I go to view the Syncthing log, it's usually empty. And the Android log doesn't show much either, and only shows things from when I open it, not from hours ago when it (presumably) tried to quit and failed.

**Comment by JohnVeness** [comment]
I don't know why this happens and I haven't looked at the code to see if you're already doing this, but maybe you can detect if the quit failed and retry, say, one minute later?

**Comment by Catfriend1** [comment]
I'll have a look, but not this week.

**Comment by prestidigitonium** [comment]
I have the same problem, it seems that settings are overwritten periodically (maybe after a sync). I've **v1.16.0.6** and **Android 10** (Crdroid 6.17), I haven't any automation or freezing app. What I noticed:
If I set the 5 min rule, it works as expected because I see the review of the current rules in the "Status" tab (main page). After some time I don't change the settings but the "Status" tab shows another set of rules, so if I uncheck and recheck the "5 min rules" then the "Status" tab is refreshed with the correct rules.

Real case (what happened):
1) I set "Works only on wifi", "Works on mobile data" and "5 min rule"
2) it works for a few hours, and the status tab shows "Works only on wifi", "Works on mobile data" and "5 min rule"
3) after a couple of day or a few hours the status page shows "Works only on wifi" (or sometimes "Works on mobile data") and no other constraints
4) I uncheck "5 min rule", I re-check "5 min rule" and the status tab shows "Works only on wifi", "Works on mobile data" and "5 min rule" again

**Comment by JohnVeness** [comment]
@prestidigitonium

In my case, my Settings don't change, as far as I've noticed. I have to do Force Stop then Follow Run Conditions to fix it (or reboot the device!), but I don't actually have to change the run condition settings.

Maybe you should enter your problem as a new issue?

**Comment by Catfriend1** [comment]
@Helium314  Does this make any clue for you? I don't know if it is a phone specific problem or old code of mine or something with the recent pr?!

**Comment by Helium314** [comment]
I looked through the parts affected by #766
When the `SyncTriggerReceiver` gets triggered by scheduled job, there is always a new job scheduled and all other jobs cancelled.

Except if Syncthing is already running and `decideShouldRun()` returns `true`. But this should not occur, because `mTimeConditionsMatch` (must be true if Syncthing is running) gets switched to false, so `decideShouldRun()` should return false.

`mTimeConditionsMatch` can be switched to true during `decideShouldRun()` if the last start/stop was more than 1 hour ago. But the stop is scheduled 5 min after start, so it's not running for that long. Although if it would already be running for too long (more than 1 houd), it would not be stopped.

One idea how Syncthing might start running for too long: in a different app I had some troubles with `AlarmManager`, which would sometimes fire alarms up to 45 min later than scheduled, even though I have doze mode completely disabled AND whitelisted this app in doze exceptions AND it's a foreground service.
If `JobScheduler` is similarly crippled by Androids dreaded power save functions (even worse for some specific manufacturers, and probably by some power save apps that have root access), then the stop job might be more than 55 min late.

If this is happening, the only solution would be to _always_ schedule a job when `SyncTriggerReceiver` receives a trigger, even though there should already be a job scheduled. Those additional stop jobs should be cancelled during normal operation, so there should be no side effects. Except for (probably) unmeasurably higher power consumption due to a few unncessary jobs... thanks to power saving?

Changing
https://github.com/Catfriend1/syncthing-android/blob/953c568c507ac993ec3581853a47bfcb4538ae0f/app/src/main/java/com/nutomic/syncthingandroid/service/RunConditionMonitor.java#L319
to
`} else { JobUtils.scheduleSyncTriggerServiceJob(context, Constants.TRIGGERED_SYNC_DURATION_SECS); }`
is the idea.

**Comment by 0xCA** [comment]
I don't think this is a network issue. After some time with this problem I decided to turn on charging only condition to avoid very significant power drain. And this setting is working perfectly, always shutting down syncthing when unplugged. I wish timer issue will be resolved soon so I can enable sync with battery too.

**Comment by Helium314** [comment]
@Catfriend1 did you try my proposed solution? Or should I do a PR?

**Comment by Catfriend1** [comment]
@prestidigitonium @JohnVeness 

Hi, could you please test @Helium314 's PR which attempts to fix the problem and let me know if the fix works for you? Here's the link to the GitHub debug build: https://github.com/Catfriend1/syncthing-android/releases/download/v1.18.1.0%2B59b9fd3c/com.github.catfriend1.syncthingandroid_github_v1.18.1.0+59b9fd3c.apk

Thanks for you follow up in advance.

Kind regards,
Catfriend1

**Comment by Helium314** [comment]
@JohnVeness @0xCA @prestidigitonium 
Did any of you try the linked debug version?
If so, does Syncthing still keep running despite being set to run on schedule?

If it still happens, I can scrap my attempt at fixing this. If not, then it might be better to wait a bit longer (that bug seems to occur rarely after all...)

**Comment by JohnVeness** [comment]
Hi, sorry, I haven't had a chance to test the debug version yet. Will do so and let you know.

**Comment by JohnVeness** [comment]
OK, I've been running this version for just over a week, and I would say that, in general, it seems to work (I haven't had to manually stop/start Syncthing), so it is an improvement, *but* the behaviour still seems imperfect to me. I'll explain.

For the first few days I noticed no problems at all. I was about to declare "problem solved", but then a couple of days ago I happened to wake my phone in the middle of the night and noticed the uptime was over an hour:
![Screenshot_20210901-010826](https://user-images.githubusercontent.com/2512915/132053969-0d47a31d-fbd8-4afc-aeb4-241b4fd4debc.png)
I waited a few minutes (while using my phone for other things), then checked again and it said the uptime was 1 minute. Presumably it had stopped then the trigger to do another 5 minute run occurred?:
![Screenshot_20210901-011706](https://user-images.githubusercontent.com/2512915/132054031-49338e1d-1bb6-4621-ac5a-d39ca764b951.png)
Checking after a few more minutes, Syncthing had stopped.

Does this behaviour sound as expected, based on the fix as written?

This morning, when waking my phone, I noticed the uptime was over 4 hours:
![Screenshot_20210903-073551](https://user-images.githubusercontent.com/2512915/132054231-6de680bf-e9c6-414e-8571-fa019242fcd1.png)
I waited a few more minutes as before, expecting it to show between 0 and 5 minutes, but it was still over 4 hours:
![Screenshot_20210903-074357](https://user-images.githubusercontent.com/2512915/132054429-23e7bc91-6f2d-435a-a4d0-9e055b6b3836.png)
I checked over an hour later, and uptime was over 5 hours:
![Screenshot_20210903-085317](https://user-images.githubusercontent.com/2512915/132054465-b258434e-2886-46fd-a595-baabc374d199.png)

At some point later in the day, I noticed Syncthing had stopped, but I couldn't tell you when. So it seems it did correct itself, but took several hours to do so.

As I say, I didn't have to manually stop/start Syncthing to get it to recover from these symptoms, which is an improvement, but it still seems the uptime can be rather excessive.

And in general, when I go into the Syncthing wrapper UI and check the Status tab, it does seem odd to the eye to see any uptime over 5 minutes. To a layman, it "feels" like at this point Syncthing should "know" that it has been running for over 5 minutes and should take immediate corrective action (e.g. stop Syncthing). To an average user, I would say it seems unintuitive and messy to ever see an uptime of over 5 minutes if the "Run according to time schedule" setting is ticked. But I guess that would need a different method to fix.

**Comment by Helium314** [comment]
> For the first few days I noticed no problems at all. I was about to declare "problem solved", but then a couple of days ago I happened to wake my phone in the middle of the night and noticed the uptime was over an hour:
> I waited a few minutes (while using my phone for other things), then checked again and it said the uptime was 1 minute. Presumably it had stopped then the trigger to do another 5 minute run occurred?:
> Checking after a few more minutes, Syncthing had stopped.
> 
> Does this behaviour sound as expected, based on the fix as written?

The (almost) immediate start of another run period is not expected.
The long run time is still possible because Android likes to delay scheduled jobs (like stopping Syncthing) by a long time if the phone is idle (e.g. at night).
But once Syncthing is stopped, the stop time is stored and a job for starting is scheduled in 1 hour.
So my first idea is that either the stop time (technically: the time of last start or stop) is not stored correctly -> should not be possible if the system clock isn't completely broken
or the start job is started way too early -> I seriously doubt that the system would start more than a few seconds earlier than scheduled.

tl;dr:
I don't know what is happening

> This morning, when waking my phone, I noticed the uptime was over 4 hours:
> I waited a few more minutes as before, expecting it to show between 0 and 5 minutes, but it was still over 4 hours:
> I checked over an hour later, and uptime was over 5 hours:
> 
> At some point later in the day, I noticed Syncthing had stopped, but I couldn't tell you when. So it seems it did correct itself, but took several hours to do so.

This fits better with my understanding: The stop job is started once the phone gets out of idle mode, but more than 1 hour after start. Since it's now more than 1 hour after last start/stop, syncthing is still in "allowed to run" condition. But the job is still scheduled in 1 hour because syncthing was supposed to be stopped.

The easiest fix for this would be to store the time when this line is called in the settings (as if it was a start/stop event)
https://github.com/Catfriend1/syncthing-android/blob/953c568c507ac993ec3581853a47bfcb4538ae0f/app/src/main/java/com/nutomic/syncthingandroid/service/RunConditionMonitor.java#L538

I will add this to my PR, and @Catfriend1 may update the test build for you.
But I somewhat doubt that this will fix everything :neutral_face:

**Comment by Helium314** [comment]
I noticed that my proposed "solution" would likely not change anything, but have another idea: while syncthing is running, jobs should always be scheduled in 5 minutes, so that it should at least stop much sooner after you notice the problem.
It's far from perfect, but at least should better than nothing...

**Comment by JohnVeness** [comment]
I wonder if it's worth considering implementing the five minute timer into the syncthing binary itself (and upstreaming it, so that laptops etc. could also use it)? As I understand it, during these long uptimes, the syncthing code *is* running, so if that checked the time rather than the Android wrapper, that would remove the vagaries of the Android job scheduling/alarm system.

**Comment by prestidigitonium** [comment]
@JohnVeness do you know the release number of a working build (before the problem)? Now I just need a working version, I get this problem every 3-4 days.

**Comment by JohnVeness** [comment]
@prestidigitonium As far as I know, there is no older version of the "5 minutes every hour" that works flawlessly. However, I have been testing a fix by @Helium314 that seems to fix it. Hopefully that will be released soon.

**Comment by 0xCA** [comment]
> @prestidigitonium As far as I know, there is no older version of the "5 minutes every hour" that works flawlessly. However, I have been testing a fix by @Helium314 that seems to fix it. Hopefully that will be released soon.

Maybe it wasn't flawlessly, but I never noticed it way back then. If it was a case, it was happening extremely rarely or not on every device. Then it was this nightmare of battery drain, and now it's fixed again.
